Update of glob-style pattern matching

riknoll opened the pull request:
This is a resubmission of #5980 that has been brought up to date with master (the old PR had a lot of non-trivial merge conflicts with the current master). The behavior differs somewhat from the specification in #5980 to better reflect the current behavior of the compiler. Excluded patterns now default to the common package directories (`node_modules`, `bower_components`, etc.) when "exclude" is not found in `tsconfig.json`. Additionally, the out directory is always excluded unless explicitly included in "files". 

@mhegazy @rbuckton @billti please take a look when you have a moment!

msftclas commented:
Hi __@riknoll__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Richard Knoll). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

riknoll commented:
@rbuckton could you take a look when you have a moment?

rbuckton commented on src/compiler/commandLineParser.ts:
```diff
@@ -849,4 +837,296 @@ namespace ts {
+                    continue;
+                }
+
+                if (IgnoreFileNamePattern.test(file)) {
```
nit: capitalization - `ignoreFileNamePattern`

rbuckton commented on src/compiler/commandLineParser.ts:
```diff
@@ -849,4 +837,296 @@ namespace ts {
+            // Remove any subpaths under an existing recursively watched directory.
+            for (const key in wildcardDirectories) {
+                if (hasProperty(wildcardDirectories, key)) {
+                    for (const recursiveKey in recursiveKeys) {
```
Should be `for..of` not `for..in`

riknoll replied:
Good catch! I've fixed this and added a test case

rbuckton commented on src/compiler/sys.ts:
```diff
@@ -380,8 +369,35 @@ namespace ts {
+                            continue;
+                        }
+                        const name = combinePaths(path, entry);
+                        const stat = _fs.statSync(name);
```
May want to put a `try..catch` around the call to `statSync`. If we don't, then we'll fail to return the contents of a directory if even one entry in the directory has more restrictive permissions that would cause `statSync` to fail.

rbuckton commented on tests/cases/unittests/cachingInServerLSHost.ts:
```diff
@@ -82,7 +82,7 @@ namespace ts {
         const rootScriptInfo = projectService.openFile(rootFile, /* openedByClient */true);
         const project = projectService.createInferredProject(rootScriptInfo);
-        project.setProjectOptions({ files: [rootScriptInfo.fileName], compilerOptions: { module: ts.ModuleKind.AMD } });
+        project.setProjectOptions( {files: [rootScriptInfo.fileName], compilerOptions: {module: ts.ModuleKind.AMD} } );
```
nit: whitespace

rbuckton commented:
Other than the comments above and the merge conflict, :+1:

riknoll commented:
@rbuckton I've taken care of the merge, addressed your comments, and added a few test cases. Would you mind checking out how I resolved the conflicts in https://github.com/Microsoft/TypeScript/pull/8841/commits/86cde9e22204be835572216b10757dcf82af8050 ?

rbuckton commented:
👍

riknoll commented:
@rbuckton thanks!

basarat commented:
wow 🎉 🌹

mhegazy commented:
//CC: @basarat, @jrieken, @blakeembrey @jbrantly, @adidahiya, @alexeagle, @martine 

this change has a breaking change to the public API, changing the arguments to `ParseConfigHost.readDirectory` and `System.readDirectory`, from
```ts
readDirectory(rootDir: string, extension: string, exclude: string[]): string[];
```
to:
```ts
readDirectory(rootDir: string, extensions: string[], excludes: string[], includes: string[]): string[];
```

Please do let us know if this is something that manageable on your side or more changes needs to be done on the TS side.

johnnyreilly commented:
Awesome!😃

martine commented:
@mhegazy thanks for thinking of me!  I didn't see any use/impl of readDirectory in Google code.

In Angular2 I found one use @alexeagle , but it's likely easy to update:
https://github.com/angular/angular/blob/3d5bb2318454ba428b816f6aee6e6aaf427d70a7/tools/%40angular/tsc-wrapped/src/tsc.ts#L51

In general I think we (Google) don't mind too much when you change APIs, especially when it's a small change that the compiler can easily spot when we update.  And usually the only reason to change an API in a large way is because it's important so we'll likely be able to make that work as well.

mhegazy commented:
thanks @martine for getting to us back on this.

texastoland commented:
I'm having trouble excluding arbitrarily deep `node_modules`. Should `**/node_modules` or `**/node_modules/**/*` do the trick?

basarat commented:
> I'm having trouble excluding arbitrarily deep node_modules.

they might still get pulled in because of `import` statements :rose:

riknoll commented:
`**/node_modules` should work
